### PR TITLE
Adding nullable to logging project

### PR DESCRIPTION
**Pull request by FuPingFranco**
Adding Nullable to Logging project

### CODE REVIEW

File: src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs [hunk 1]
```diff
@@ -265,7 +265,7 @@ public void WriteCritical(string message, params object[] args)
         [NonEvent]
         public void Write(EventLevel level, Exception innerException, string message)
         {
-            Write(level, innerException, message, null);
+            Write(level, innerException, message, null!);
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs]: `null!` hints that something isn't right here.

Why did we end up with this solution?

File: src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs [hunk 2]
```diff
@@ -291,7 +291,7 @@ public void Write(EventLevel level, Exception innerException, string message, pa
             if (!HeaderWritten)
             {
                 // Obtain the current library version dynamically.
-                WriteAlways(string.Format(CultureInfo.InvariantCulture, _versionLogMessage, typeof(IdentityModelEventSource).GetTypeInfo().Assembly.GetName().Version.ToString()));
+                WriteAlways(string.Format(CultureInfo.InvariantCulture, _versionLogMessage, typeof(IdentityModelEventSource).GetTypeInfo().Assembly!.GetName().Version!.ToString()));
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs]: I don't understand, why is Assembly! and Version! required here?

File: src/Microsoft.IdentityModel.Logging/IdentityModelTelemetryUtil.cs [hunk 3]
```diff
@@ -44,7 +44,7 @@ public static class IdentityModelTelemetryUtil
         /// <summary>
         /// Get the string that represents the client version.
         /// </summary>
-        public static string ClientVer => typeof(IdentityModelTelemetryUtil).GetTypeInfo().Assembly.GetName().Version.ToString();
+        public static string ClientVer => typeof(IdentityModelTelemetryUtil).GetTypeInfo().Assembly!.GetName().Version!.ToString();
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/IdentityModelTelemetryUtil.cs]: same comment, why is this required?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs [hunk 4]
```diff
@@ -262,10 +262,10 @@ public static Exception LogExceptionMessage(Exception exception)
         public static Exception LogExceptionMessage(EventLevel eventLevel, Exception exception)
         {
             if (exception == null)
-                return null;
+                return null!;
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/LogHelper.cs]: this doesn't seem right, why not change the return type to `Exception?`

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs [hunk 5]
```diff
@@ -262,10 +262,10 @@ public static Exception LogExceptionMessage(Exception exception)
...
 
             if (IdentityModelEventSource.Logger.IsEnabled(eventLevel, EventKeywords.All))
-                IdentityModelEventSource.Logger.Write(eventLevel, exception.InnerException, exception.Message);
+                IdentityModelEventSource.Logger.Write(eventLevel, exception.InnerException!, exception.Message);
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/LogHelper.cs]: how do we know innerexception is not null here?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs [hunk 6]
```diff
@@ -333,22 +333,22 @@ public static void LogWarning(string message, params object[] args)
                 message = format;
 
             if (IdentityModelEventSource.Logger.IsEnabled(eventLevel, EventKeywords.All))
-                IdentityModelEventSource.Logger.Write(eventLevel, innerException, message);
+                IdentityModelEventSource.Logger.Write(eventLevel, innerException!, message);
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/LogHelper.cs]: same question, how do we know innerException is not null?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs [hunk 7]
```diff
@@ -333,22 +333,22 @@ public static void LogWarning(string message, params object[] args)
...
 
             EventLogLevel eventLogLevel = EventLevelToEventLogLevel(eventLevel);
             if (Logger.IsEnabled(eventLogLevel))
                 Logger.Log(WriteEntry(eventLogLevel, innerException, message, null));
 
             if (innerException != null) 
                 if (string.IsNullOrEmpty(argumentName))
-                    return (T)Activator.CreateInstance(typeof(T), message, innerException);
+                    return (T)Activator.CreateInstance(typeof(T), message, innerException)!;
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/LogHelper.cs]: for all of these, Activator can return null though, why are we suppressing that here?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs [hunk 8]
```diff
@@ -383,12 +383,12 @@ private static object SanitizeSecurityArtifact(object arg)
...
 
             if (IdentityModelEventSource.LogCompleteSecurityArtifact && arg is ISafeLogSecurityArtifact)
-                return (arg as ISafeLogSecurityArtifact).UnsafeToString();
+                return ((ISafeLogSecurityArtifact)arg).UnsafeToString();
 
             return arg;
         }
 
-        private static string RemovePII(object arg)
+        private static string? RemovePII(object arg)
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/LogHelper.cs]: this doesn't look like it can return null?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs [hunk 9]
```diff
@@ -401,7 +401,7 @@ private static string RemovePII(object arg)
 
         internal static bool IsCustomException(Exception ex)
         {
-            return ex.GetType().FullName.StartsWith("Microsoft.IdentityModel.", StringComparison.Ordinal);
+            return ex.GetType().FullName!.StartsWith("Microsoft.IdentityModel.", StringComparison.Ordinal);
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/LogHelper.cs:414]: why is this needed?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs [hunk 10]
```diff
@@ -436,10 +436,10 @@ public static object MarkAsSecurityArtifact(object arg, Func<object, string> cal
         /// <param name="innerException"><see cref="Exception"/></param>
         /// <param name="message">The log message.</param>
         /// <param name="args">An object array that contains zero or more objects to format.</param>
-        private static LogEntry WriteEntry(EventLogLevel eventLogLevel, Exception innerException, string message, params object[] args)
+        private static LogEntry WriteEntry(EventLogLevel eventLogLevel, Exception? innerException, string message, params object[]? args)
         {
             if (string.IsNullOrEmpty(message))
-                return null;
+                return null!;
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/LogHelper.cs]: I don't understand

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs [hunk 11]
```diff
@@ -458,8 +458,8 @@ private static LogEntry WriteEntry(EventLogLevel eventLogLevel, Exception innerE
             // Prefix header (library version, DateTime, whether PII is ON/OFF) to the first message logged by Wilson.
             if (!_isHeaderWritten)
             {
-                string headerMessage = string.Format(CultureInfo.InvariantCulture, "Microsoft.IdentityModel Version: {0}. Date {1}. {2}",
-                    typeof(IdentityModelEventSource).Assembly.GetName().Version.ToString(),
+                string headerMessage = string.Format(CultureInfo.InvariantCulture!, "Microsoft.IdentityModel Version: {0}. Date {1}. {2}",
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/LogHelper.cs]: why is this needed?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs [hunk 12]
```diff
@@ -458,8 +458,8 @@ private static LogEntry WriteEntry(EventLogLevel eventLogLevel, Exception innerE
...
+                    typeof(IdentityModelEventSource).Assembly!.GetName().Version!.ToString(),
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/LogHelper.cs]: why are these needed?

File: src/Microsoft.IdentityModel.Logging/TextWriterEventListener.cs [hunk 13]
```diff
@@ -91,7 +91,7 @@ protected override void OnEventWritten(EventWrittenEventArgs eventData)
 
             for (int i = 0; i < eventData.Payload.Count; i++)
             {
-                _streamWriter.WriteLine(eventData.Payload[i].ToString());
+                _streamWriter.WriteLine(eventData.Payload[i]!.ToString());
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/TextWriterEventListener.cs:96]: is this safe?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 6 above)
Comment (jmprieur) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: the change is rather in Logger.Write to accept null

File: src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs (hunk 1 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs]: I'd think that you would need:
`public void Write(EventLevel level, Exception innerException, string message, params object[]? args)` 
and then 
`public void WriteAlways(string message, params object[]? args)`

which already has a null check for null `args`

File: src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs (hunk 2 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs]: I'm not sure `!` is needed for either. Do you get warnings without?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 4 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: You need to return null here, so adding `null!` is not correct because we are returning null. Only add `!` when you know for sure the value will never be null.

Instead, note that the return type can be null instead:
`public static Exception? LogExceptionMessage(EventLevel eventLevel, Exception exception)`

and bubble it up:
`public static Exception? LogExceptionMessage(Exception exception)`

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 5 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: +1, also, I don't think this is needed.

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 6 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: I don't think so. The args can be null, but I don't think we'll get any nullable warnings with the code as-is.

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 7 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: I don't think `!` is needed on any of them.

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 8 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: Correct...why was this added as a possible null return?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 10 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: The issue, I think, is this will return null, so adding a null return type `LogEntry?` means things like this 
```csharp
if (Logger.IsEnabled(eventLogLevel))
                Logger.Log(WriteEntry(eventLogLevel, exception.InnerException, exception.Message, null));
``` 
could have null references.

I'm wondering why we return null here anyway if message is empty? We could remove the null check on the message, looks like the only place where it's not passed is here, and we can check for null before:
`Logger.Log(WriteEntry(eventLogLevel, exception.InnerException, exception.Message ?? string.Empty, null));` and pass in an empty string instead. 

Not sure the right thing to do here, but it's a bit tricky.

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 11 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: +1 should not be needed

File: src/Microsoft.IdentityModel.Logging/TextWriterEventListener.cs (hunk 13 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/TextWriterEventListener.cs:96]: why do you think this is needed @FuPingFranco ? Are you getting a warning in the IDE?

**Review by jennyf19 — CHANGES_REQUESTED** [review]
left a lot of questions and some suggestions. The nullables do not look to be added in the correct places.

File: src/Microsoft.IdentityModel.Logging/TextWriterEventListener.cs (hunk 13 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/TextWriterEventListener.cs:96]: @keegan-caruso I believe it is safe, null checks for  eventData and Payload are checked on lines 83 and 86 respectively.
@jennyf19  Yes, I get warnings for error code CS8602-Dereference of a possibly null reference. (happens when we set nullables but it doesn't know we already performed null checks previous to using it this way)

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 12 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: IDE throws error code CS8602-Dereference of a possibly null reference. Version is not nullable in System.Reflection so it causes noise

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 11 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: It was throwing an error before but seems to not be the case anymore. Will remove thanks!

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 9 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs:414]: IDE throws error code CS8602-Dereference of a possibly null reference. FullName is not nullable in System so it causes noise

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 8 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: It used in line 375 as a return value

 if (!IdentityModelEventSource.ShowPII)
                return string.Format(CultureInfo.InvariantCulture, format, args.Select(RemovePII).ToArray()); 

IDE throws error code CS8603-Possible null reference return

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 7 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: @keegan-caruso Yes Activator can return null because it's been handled, but IDE throws error code CS8603-Possible null reference return.

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 6 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: @jmprieur  Is correct, I took off the warning during development to deal with another error and forgot to take it off to add the nullable arg to the logger.Write

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 5 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: Same as the next comment

File: src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs (hunk 2 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs]: @jennyf19  Yeap, without it, IDE throws error code CS8602-Dereference of a possibly null reference. Version is not nullable in System.Reflection so it causes noise

File: src/Microsoft.IdentityModel.Logging/IdentityModelTelemetryUtil.cs (hunk 3 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/IdentityModelTelemetryUtil.cs]: IDE throws error code CS8602-Dereference of a possibly null reference. Version is not nullable in System.Reflection so it causes noise

File: src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs (hunk 1 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs]: Great catch, I muted off this error because I started working on a different as it was very repetitive from lines 306-328. I'll update this.

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 4 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: I agree with returning Exception? but it leads to error code (CS8597- Thrown value may be null )
in all instances where we have something like "throw LogExceptionMessage(new WsFederationException(LogMessages.IDX22902)"

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 10 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: @jennyf19 that is a solution however, an empty string and a Null string are not equal so, I opted to keep the current flow by using !, not the best solution or a solution at all but, given the deadline I opted for keeping the current flow untouched.

File: src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs (hunk 1 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs]: Can you resolve the comments which are fixed? thank you

File: src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs (hunk 2 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/IdentityModelEventSource.cs]: I had Build on, not Build + IntelliSense, so I didn't see this one. You can actually do `GetTypeInfo().Assembly.GetName().Version?.ToString()));` instead

File: src/Microsoft.IdentityModel.Logging/IdentityModelTelemetryUtil.cs (hunk 3 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/IdentityModelTelemetryUtil.cs]: should be `GetTypeInfo().Assembly.GetName().Version?.ToString();` instead

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 7 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: should be `return (T?)Activator.CreateInstance(typeof(T), message);`

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 8 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: should be `return arg.ToString()!; // the null check is done in FormatInvariant`

File: src/Microsoft.IdentityModel.Logging/TextWriterEventListener.cs (hunk 13 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/TextWriterEventListener.cs:96]: should be `_streamWriter.WriteLine(eventData!.Payload[i]?.ToString());`

**Comment by jennyf19** [comment]
> left a lot of questions and some suggestions. The nullables do not look to be added in the correct places.

@FuPingFranco I think addressed most of the comments in this [PR](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2215), so please take a look, merge when you want and resolve comments which are addressed. thanks!

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 5 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: Made arg nullable

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 6 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: Made arg nullable

File: src/Microsoft.IdentityModel.Logging/IdentityModelTelemetryUtil.cs (hunk 3 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/IdentityModelTelemetryUtil.cs]: Changed implementation, pls review again for correctness. Thanks!

File: src/Microsoft.IdentityModel.Logging/TextWriterEventListener.cs (hunk 13 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/TextWriterEventListener.cs:96]: @jennyf19 added condition instead as per Marcin recommendation.

File: src/Microsoft.IdentityModel.Logging/IdentityModelTelemetryUtil.cs (hunk 3 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/IdentityModelTelemetryUtil.cs]: I would declare this as null: `public static string? ClientVer => ...` and then earlier handle the null by passing an empty string. We wouldn't want to mess up with the values sent to eSTS. Maybe something like `[versionTelemetry] = ClientVer ?? string.Empy`

Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/IdentityModelTelemetryUtil.cs]: Makes sense, I updated it to reflect your suggestion. Thanks Jenny!

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 10 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: can we do `message = string.Empty;`?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 4 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: why not `exception = exception ?? throw new ArgumentNullException(nameof(exception));` for example?

**Comment by jennyf19** [comment]
>         public static string FormatInvariant(string format, params object[] args)

I'm surprised you don't need `string?` here as format can be null. 

---
Refers to: src/Microsoft.IdentityModel.Logging/LogHelper.cs:366 in ed96099. [](commit_id = ed960993b38122ecbd186eed8b764c8ddb9fe26f, deletion_comment = False)

**Comment by jennyf19** [comment]
Left a few comments. still needs a bit of work.

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 4 above)
Comment (keegan-caruso) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: returning null is still incorrect.

I would argue that passing null is a usage error of the API.

It would be a behavioral breaking change, but what about throwing an argument null exception if passed null? We would need to include it in the breaking change list, but it is targeting the 7x branch.

Thoughts @jennyf19  and @FuPingFranco ?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 7 above)
Comment (keegan-caruso) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: this is still incorrect. We either need to actually check that the object created is not null, or we need to change the return type to T?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 10 above)
Comment (keegan-caruso) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: this one still needs to be figured out, we can't return null here.

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 7 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: Is this still an issue @FuPingFranco ?

Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: @jennyf19 Yes, this method cannot return null as is right now. For some reason when you help me look into it with the outdated branch, it was returning nulls then as I see it in this [PR](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2215).

@keegan-caruso I like the idea of checking for nulls here more than changing the return type but, not sure what would be appropriate return/throw if it is null. Any thoughts?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 4 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: @keegan-caruso i agree with you about return null, that seems to be a programming mistake and can lead to a null ref at any time.

How about we add a debug assert if the exception passed in is null and return new ArgumentNullException("null exception was passed to logger");

@jennyf19  not sure where you are suggesting we have this code.
why not exception = exception ?? throw new ArgumentNullException(nameof(exception)); for example?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 7 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: @jennyf19 After synching with Keegan, we changed the implementation slightly which took care of this. Resolving this comment Thank you both!

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 10 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: Will be retuning new ArgumentNullException and adding it to the breaking changes. Resolving this Thread. Thank you both!

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 4 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: I agree, I have made the change to thow an ArgumentNullException post null check and will include it to the coming breaking changes. Resolving this thread, thanks everyone!

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs [hunk 14]
```diff
@@ -383,7 +393,7 @@ private static object SanitizeSecurityArtifact(object arg)
                 return "null";
 
             if (IdentityModelEventSource.LogCompleteSecurityArtifact && arg is ISafeLogSecurityArtifact)
-                return (arg as ISafeLogSecurityArtifact).UnsafeToString();
+                return ((ISafeLogSecurityArtifact)arg).UnsafeToString();
```
Comment (jmprieur) [inline src/Microsoft.IdentityModel.Logging/LogHelper.cs]: why this change?

**Review by jmprieur — APPROVED** [review]
LGTM
I just left a question

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 14 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: @jmprieur Using 'as' leads to error CS8602- Dereference of a possibly null reference after we enable nullables. This is because when using 'as' it will never return an InvalidCastExeception if it happens, instead it will ansign and return null which is not right in this case. An alternative is to declare arg as a nullable which is still not correct as we never return null from this method as shown in line 393 where we return a string named "null" if the argument is ever null.

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs [hunk 15]
```diff
@@ -339,16 +339,26 @@ public static void LogWarning(string message, params object[] args)
...
                 if (string.IsNullOrEmpty(argumentName))
-                    return (T)Activator.CreateInstance(typeof(T), message);
+                    value = Activator.CreateInstance(typeof(T), message) as T;
                 else
-                    return (T)Activator.CreateInstance(typeof(T), argumentName, message);
+                    value = Activator.CreateInstance(typeof(T), argumentName, message) as T;
+            }
+
+            if(value is null)
+                throw new ArgumentException("Newly created Exception is null");
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/LogHelper.cs:359]: we need a better error message, this isn't actionable.

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs [hunk 16]
```diff
@@ -262,7 +262,7 @@ public static Exception LogExceptionMessage(Exception exception)
         public static Exception LogExceptionMessage(EventLevel eventLevel, Exception exception)
         {
             if (exception == null)
-                return null;
+                throw new ArgumentNullException(nameof(exception), "null exception was passed to logger");
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/LogHelper.cs]: does the message add any value in this case? or can we just pass the param name?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 14 above)
Comment (keegan-caruso) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: suggestion: use pattern matching to clean this up

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 16 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: @keegan-caruso We could just pass the param name, but the I think the message would help more quickly identify why and where this ArgumentNullException happened.

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 15 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs:359]: @keegan-caruso I can elaborate more on this error message however, because 'value' is created local to this method, there isn't a specific param or inner exception I can include as part of the ArgumentException. Do you have any suggestions to improve on this?

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 16 above)
Comment (keegan-caruso) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: "why" is irrelevant in this case. It is a ANE, it should be self-describing.

"where" should be clear from a stack trace. If we need more we should use an IDX prefaced log message.

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs [hunk 17]
```diff
@@ -436,10 +446,10 @@ public static object MarkAsSecurityArtifact(object arg, Func<object, string> cal
         /// <param name="innerException"><see cref="Exception"/></param>
         /// <param name="message">The log message.</param>
         /// <param name="args">An object array that contains zero or more objects to format.</param>
-        private static LogEntry WriteEntry(EventLogLevel eventLogLevel, Exception innerException, string message, params object[] args)
+        private static LogEntry WriteEntry(EventLogLevel eventLogLevel, Exception? innerException, string message, params object[]? args)
         {
             if (string.IsNullOrEmpty(message))
-                return null;
+                throw new ArgumentNullException(nameof(message), "null message was passed to new entry");
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Logging/LogHelper.cs]: same comment as above

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 15 above)
Comment (keegan-caruso) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs:359]: either build up the message based on the path taken above or list out all of the params even if they are potentially null/unused? (while adhering to PII awareness)

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 16 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: Fair enough, I will remove the exception message. Thanks Keegan!

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 17 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs]: Same reply as above.

File: src/Microsoft.IdentityModel.Logging/LogHelper.cs (hunk 15 above)
Comment (keegan-caruso) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs:359]: at a minimum we could at least call out the type?

Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Logging/LogHelper.cs:359]: I am not sure which one will make the most sense in here. @jennyf19 @jmprieur do you any suggestions on improving this message?

**Comment by jennyf19** [comment]
As discussed, closing this PR as we are on a tight ship schedule.
